File: 0113-path-sum-ii/0113-path-sum-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    result = list()
    path = list()
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        self.result = list()
        self.path = list()
        self.target = targetSum
        self.helper(root, 0)
        return self.result
    
    def helper(self, root, currSum):
        if root is None:
            return
        
        self.path.append(root.val)
        currSum += root.val

        if not root.left and not root.right and self.target == currSum:
            self.result.append(self.path.copy())

        self.helper(root.left, currSum)
        self.helper(root.right, currSum)
        # print(self.path, "Before")
        self.path.pop()
        # print(self.path, "After")

File: 0113-path-sum-ii/0113-path-sum-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<List<Integer>> result;
    List<Integer> path;
    Integer target;
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        this.result = new ArrayList<>();
        this.path = new ArrayList<>();
        this.target = targetSum;
        
        helper(root, 0);

        return result;
    }

    private void helper(TreeNode root, int pathSum){
        //base
        if(root == null){
            return;
        }
        
        path.add(root.val);
        pathSum += root.val;


        if(root.left == null && root.right == null && target == pathSum){
            
            result.add(new ArrayList<>(path));
        }
        

        helper(root.left, pathSum);
   
        
        helper(root.right, pathSum);
        path.removeLast();

    }
}